// envio@2.26.0/src/bindings/Hrtime.res.js
// Generated by ReScript, PLEASE EDIT WITH CARE
'use strict';


function nanoToMilli(nano) {
  return nano / 1000000;
}

function secToMilli(sec) {
  return sec * 1000;
}

function nanoToTimeTuple(nano) {
  var seconds = Math.floor(nano / 1000000000);
  var nanos = nano % 1000000000;
  return [
          seconds,
          nanos
        ];
}

function timeElapsedToNewRef(elapsed, ref) {
  var match = nanoToTimeTuple(elapsed[1] + ref[1]);
  return [
          elapsed[0] + ref[0] + match[0],
          match[1]
        ];
}

function toMillis(param) {
  return param[0] * 1000 + param[1] / 1000000;
}

function toInt($$float) {
  return $$float | 0;
}

function floatFromMillis(prim) {
  return prim;
}

function makeTimer(prim) {
  return process.hrtime();
}

function timeSince(prim) {
  return process.hrtime(prim);
}

var intFromMillis = toInt;

var intFromNanos = toInt;

var intFromSeconds = toInt;

exports.makeTimer = makeTimer;
exports.timeSince = timeSince;
exports.nanoToMilli = nanoToMilli;
exports.secToMilli = secToMilli;
exports.timeElapsedToNewRef = timeElapsedToNewRef;
exports.toMillis = toMillis;
exports.intFromMillis = intFromMillis;
exports.intFromNanos = intFromNanos;
exports.intFromSeconds = intFromSeconds;
exports.floatFromMillis = floatFromMillis;
/* No side effect */
